controller/resource.go

```diff
@@ -146,48 +179,209 @@ func (s ServiceStateLister) fetchECSTasks() (map[string]struct{}, error) {
+				}
+			}
+		}
+
+		tokenList, _, err := s.ConsulClient.ACL().TokenList(opts)
+		if err != nil {
+			return aclState, err
+		}
+
+		for _, token := range tokenList {
+			//if isInCluster(s.Cluster, token.Description) && len(token.Policies) == 1 {
+			if isInCluster(s.Cluster, token.Description) {
+				qname := serviceNameFromDescription(token.Description)
+				if state, ok := aclState[qname]; ok {
+					state.ACLTokens = append(state.ACLTokens, token)
+				} else {
+					aclState[qname] = &ServiceState{ACLTokens: []*api.ACLTokenListEntry{token}}
+				}
 			}
 		}
 	}
 
-	return aclTokens, nil
+	return aclState, nil
+}
+
+// prepareNamespaces ensures that for every service in the cluster the namespace
+// exists and the cross-partition/cross-namespace read policy exists.
+func (s ServiceStateLister) ReconcileNamespaces(resources []Resource) error {
+	if !PartitionsEnabled(s.Partition) {
+		return nil
+	}
+
+	// create the cross-namespace read policy
+	if err := s.upsertCrossNSPolicy(); err != nil {
+		return err
+	}
+
+	// retrieve the list of unique namespaces from the map of services
+	ns := uniqueNamespaces(resources)
+	if len(ns) < 2 {
+		// only the default namespace exists, so nothing to do
+		return nil
+	}
+
+	if err := s.createNamespaces(ns); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// upsertCrossNSPolicy creates the cross-namespace read policy in the local
+// partition and default namespace, if it does not already exist.
+func (s ServiceStateLister) upsertCrossNSPolicy() error {
+	policy, _, err := s.ConsulClient.ACL().PolicyReadByName(
+		xnsPolicyName,
+		&api.QueryOptions{Partition: s.Partition, Namespace: DefaultNamespace},
+	)
+	if err != nil && !IsACLNotFoundError(err) {
+		return fmt.Errorf("reading cross-namespace policy: %w", err)
+	}
+
+	if policy != nil {
+		// the policy already exists so return.
+		return nil
+	}
+
+	// create the policy in the local partition and default namespace.
+	_, _, err = s.ConsulClient.ACL().PolicyCreate(&api.ACLPolicy{
+		Name:        xnsPolicyName,
+		Description: xnsPolicyDesc,
+		Partition:   s.Partition,
+		Namespace:   DefaultNamespace,
+		Rules:       fmt.Sprintf(xnsPolicyTpl, s.Partition),
+	}, nil)
+	if err != nil {
+		return fmt.Errorf("creating cross-namespace policy: %w", err)
+	}
+
+	s.Log.Info("created cross-namespace policy", xnsPolicyName)
+	return nil
+}
+
+// createNamespaces creates each namespace in the list.
+// It also attaches the cross-namespace read policy to the namespace so that
+// each token created in the namespace inherits that policy.
+//
+// It does nothing for namespaces that already exist.
+func (s ServiceStateLister) createNamespaces(ns []string) error {
+	// retrieve the list of existing namespaces
+	existingNS, _, err := s.ConsulClient.Namespaces().List(&api.QueryOptions{Partition: s.Partition})
+	if err != nil {
+		return err
+	}
+
+	// create any namespaces that do not already exist
+	create := make([]string, 0, len(ns))
+	for _, n := range ns {
```

Comment:
nit: Related to my other comment, if `uniqueNamespaces` returns a set (i.e. `map[string]struct{}`) and this accepts that set, then can we avoid this double nested loop? It might be slightly awkward and not matter much though.

Reply:
Yes, that was a really good call. I think that using a set has made this function a bit more palatable. Using the `go-multierror` also helped clean up the code.